gems/aws-sdk-core/lib/aws-sdk-core/instance_profile_credentials.rb

```diff
@@ -97,260 +76,128 @@ def initialize(options = {})
-            if @expiration && @expiration < Time.now
-              @no_refresh_until = Time.now + refresh_offset
-              warn_expired_credentials
-            end
-          else
-            #  credentials are already set, update them only if the new ones are not empty
-            if !c['AccessKeyId'] || c['AccessKeyId'].empty?
-              # error getting new credentials
-              @no_refresh_until = Time.now + refresh_offset
-              warn_expired_credentials
-            else
-              @credentials = Credentials.new(
-                c['AccessKeyId'],
-                c['SecretAccessKey'],
-                c['Token']
-              )
-              @expiration = c['Expiration'] ? Time.iso8601(c['Expiration']) : nil
-              if @expiration && @expiration < Time.now
-                @no_refresh_until = Time.now + refresh_offset
-                warn_expired_credentials
-              end
-            end
+      new_creds =
+        begin
+          retry_errors([Aws::Json::ParseError]) do
+            Aws::Json.load(fetch_credentials)
           end
+        rescue Aws::Json::ParseError
+          raise Aws::Errors::MetadataParserError
         end
-      rescue Aws::Json::ParseError
-        raise Aws::Errors::MetadataParserError
-      end
-    end
 
-    def get_credentials
-      # Retry loading credentials a configurable number of times if
-      # the instance metadata service is not responding.
-      if _metadata_disabled?
-        '{}'
+      if !empty_credentials?(@credentials) &&
+         (!new_creds['AccessKeyId'] || new_creds['AccessKeyId'].empty?)
+        @no_refresh_until = Time.now + refresh_offset
+        warn_expired_credentials
       else
-        begin
-          retry_errors(NETWORK_ERRORS, max_retries: @retries) do
-            open_connection do |conn|
-              # attempt to fetch token to start secure flow first
-              # and rescue to failover
-              fetch_token(conn) unless @imds_v1_fallback
-              token = @token.value if token_set?
-
-              # disable insecure flow if we couldn't get token
-              # and imds v1 is disabled
-              raise TokenRetrivalError if token.nil? && @disable_imds_v1
-
-              _get_credentials(conn, token)
-            end
-          end
-        rescue => e
-          warn("Error retrieving instance profile credentials: #{e}")
-          '{}'
-        end
+        update_credentials(new_creds)
       end
     end
 
-    def fetch_token(conn)
-      retry_errors(NETWORK_ERRORS, max_retries: @retries) do
-        unless token_set?
-          created_time = Time.now
-          token_value, ttl = http_put(
-            conn, METADATA_TOKEN_PATH, @token_ttl
-          )
-          @token = Token.new(token_value, ttl, created_time) if token_value && ttl
+    def fetch_credentials
+      return '{}' if @disable_ec2_metadata
+
+      profile_name =
```

Comment:
I think fetching profile name should be a separate method for readability - it can return the instance variable if set, otherwise it should fetch it using the client, and then cache it as the instance variable. Right now, you'll do this check every time, and always fetch a profile name. I believe it should be set, and then only cleared when there is a 404. Check the SEP guidance regarding resolvedProfile.

Reply:
I meant to refactor this bit better to match SEP guidance (also need to include 2.1 updates)